From 1c53281831611837de4358b6e75de594867f46b0 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Mon, 9 Jul 2007 19:37:37 +0000
Subject: [PATCH] CONFIGURATION-282: Initialize default expression engine for
 HierarchicalConfiguration lazily if it is null. This should avoid NPEs after
 redeployment.

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/configuration/trunk@554746 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration/HierarchicalConfiguration.java       | 10 +++++++---
 xdocs/changes.xml                                      |  5 +++++
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java b/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java
index 6790dd0308..ad839b9440 100644
--- a/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java
+++ b/src/java/org/apache/commons/configuration/HierarchicalConfiguration.java
@@ -151,7 +151,7 @@ public class HierarchicalConfiguration extends AbstractConfiguration implements
     private static final long serialVersionUID = 3373812230395363192L;
 
     /** Stores the default expression engine to be used for new objects.*/
-    private static ExpressionEngine defaultExpressionEngine = new DefaultExpressionEngine();
+    private static ExpressionEngine defaultExpressionEngine;
 
     /** Stores the root node of this configuration. This field is required for
      * backwards compatibility only.
@@ -258,8 +258,12 @@ public void setRootNode(ConfigurationNode rootNode)
      * @return the default expression engine
      * @since 1.3
      */
-    public static ExpressionEngine getDefaultExpressionEngine()
+    public static synchronized ExpressionEngine getDefaultExpressionEngine()
     {
+        if (defaultExpressionEngine == null)
+        {
+            defaultExpressionEngine = new DefaultExpressionEngine();
+        }
         return defaultExpressionEngine;
     }
 
@@ -272,7 +276,7 @@ public static ExpressionEngine getDefaultExpressionEngine()
      * @param engine the new default expression engine
      * @since 1.3
      */
-    public static void setDefaultExpressionEngine(ExpressionEngine engine)
+    public static synchronized void setDefaultExpressionEngine(ExpressionEngine engine)
     {
         if (engine == null)
         {
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index bd0300f7ad..85b0ec1bb2 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,11 @@
 
   <body>
     <release version="1.5-SNAPSHOT" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-282">
+        The default expression engine used by HierarchicalConfiguration
+        instances is now lazily initialized. This avoids NullPointerExceptions
+        in certain server environments after a redeploy.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-281">
         Cycles in the JNDI tree no longer cause a stack overflow in
         JNDIConfiguration.
